### PR TITLE
[8.x] Added sole() and soleWhere() methods for Collections

### PR DESCRIPTION
**Pull request by ash-jc-allen**
Hey! This PR should hopefully add new ` sole() ` and ` soleWhere() ` methods to Collections and LazyCollections. I've taken inspiration from the similar ` sole() ` method for the database queries and think that these could be pretty handy additions.

The ` sole() ` method should work in a similar way to the ` first() ` method does but does not require a default value as a second parameter. Likewise, the ` soleWhere() ` method should work in a similar way to the ` firstWhere() ` method. 

If either of the methods are executed and there are no items to be returned, an ` Illuminate\Collections\ItemNotFoundException ` will be thrown. If the either of the methods are executed are there is more than one item to be returned, an ` Illuminate\Collections\MultipleItemsFoundException ` will be thrown.

Here's a couple of quick examples of how they could be used:

**sole()**:

```php
$collection = collect([
    ['name' => 'foo'],
    ['name' => 'bar'],
    ['name' => 'bar'],
]);

// $result will be equal to: ['name' => 'foo']
$result = $collection->where('name', 'foo')->sole();

// $result will be equal to: ['name' => 'foo']
$result = $collection->sole(function ($value) {
    return $value['name'] === 'foo';
});

// This will throw an ItemNotFoundException
$collection->where('name', 'INVALID')->sole();

// This will throw a MultipleItemsFoundException
$collection->where('name', 'bar')->sole();
```

**soleWhere()**:
```php
$collection = collect([
    ['name' => 'foo'],
    ['name' => 'bar'],
    ['name' => 'bar'],
]);

// $result will be equal to: ['name' => 'foo']
$result = $collection->soleWhere('name', 'foo');

// This will throw an ItemNotFoundException
$collection->soleWhere('name', 'INVALID');

// This will throw a MultipleItemsFoundException
$collection->soleWhere('name', 'bar');
```
If there's anything that might need changing on these, please let me know :smile:

### CODE REVIEW

**Comment by driesvints** [comment]
Ping @JosephSilber

File: src/Illuminate/Collections/Collection.php [hunk 1]
```diff
@@ -1050,6 +1052,31 @@ public function splitIn($numberOfGroups)
...
 
+    /**
+     * Get the first item in the collection, but only if exactly
+     * item exists. Otherwise, throw an exception.
+     *
+     * @param  callable|null  $callback
+     * @return mixed
+     *
+     * @throws ItemNotFoundException
+     * @throws MultipleItemsFoundException
```
Comment (JosephSilber) [inline src/Illuminate/Collections/Collection.php:1063]: We always use fully qualified class names in DocBlocks.

File: src/Illuminate/Collections/Enumerable.php [hunk 2]
```diff
@@ -808,6 +810,30 @@ public function slice($offset, $length = null);
...
+     * Get the first item by the given key value pair, but only if
+     * exactly one item matches the criteria. Otherwise, throw
+     * an exception.
+     *
+     * @param  string  $key
+     * @param  mixed  $operator
+     * @param  mixed  $value
+     * @return mixed
+     */
+    public function soleWhere($key, $operator = null, $value = null);
```
Comment (JosephSilber) [inline src/Illuminate/Collections/Enumerable.php:835]: We can't add methods to an interface mid-cycle.

Remove these from the interface. After it's merged, submit a separate PR to `master` to add them to the interface.

File: src/Illuminate/Collections/LazyCollection.php [hunk 3]
```diff
@@ -1010,6 +1012,51 @@ public function split($numberOfGroups)
...
+     * Get the first item in the collection, but only if exactly
+     * item exists. Otherwise, throw an exception.
+     *
+     * @param  callable|null  $callback
+     * @return mixed
+     *
+     * @throws ItemNotFoundException
+     * @throws MultipleItemsFoundException
+     */
+    public function sole(callable $callback = null)
```
Comment (JosephSilber) [inline src/Illuminate/Collections/LazyCollection.php:1025]: This implementation is way more complex than it needs to be.

Something like this (not tested) should suffice:

```php
public function sole(callable $callback = null)
{
    $items = $this
        ->when($callback)
        ->filter($callback)
        ->take(2)
        ->collect();

    if ($items->isEmpty()) {
        throw new ItemNotFoundException;
    }

    if (! $items->containsOneItem()) {
        throw new MultipleItemsFoundException;
    }

    return $items[0];
}
```

File: tests/Support/SupportCollectionTest.php [hunk 4]
```diff
@@ -66,6 +68,90 @@ public function testFirstWithDefaultAndWithoutCallback($collection)
         $this->assertSame('default', $result);
     }
 
+    /**
+     * @dataProvider collectionClassProvider
+     */
+    public function testSoleReturnsFirstItemInCollectionIfOnlyOneExists($collection)
```
Comment (JosephSilber) [inline tests/Support/SupportCollectionTest.php:74]: Please also add tests in [SupportLazyCollectionIsLazyTest.php](https://github.com/laravel/framework/blob/8.x/tests/Support/SupportLazyCollectionIsLazyTest.php) ensuring this method is lazy.

File: src/Illuminate/Collections/Traits/EnumeratesValues.php [hunk 5]
```diff
@@ -453,6 +453,21 @@ public function sum($callback = null)
...
+     *
+     * @param  string  $key
+     * @param  mixed  $operator
+     * @param  mixed  $value
+     * @return mixed
+     */
+    public function soleWhere($key, $operator = null, $value = null)
+    {
+        return $this->sole($this->operatorForWhere(...func_get_args()));
+    }
```
Comment (JosephSilber) [inline src/Illuminate/Collections/Traits/EnumeratesValues.php:469]: Since, unlike the `first` method, `sole` doesn't take a `$default` parameter , there's no need for a separate `soleWhere` method. Constraints can be passed into the `sole` method directly, like `contains`.

File: src/Illuminate/Collections/Collection.php [hunk 6]
```diff
@@ -1050,6 +1052,31 @@ public function splitIn($numberOfGroups)
...
+     */
+    public function sole(callable $callback = null)
+    {
+        $items = $this->filter($callback);
```
Comment (JosephSilber) [inline src/Illuminate/Collections/Collection.php:1067]: The code as is will filter out `null`s when no `$callback` has been provided.

```suggestion
        $items = $this->when($callback)->filter($callback);
```

**Comment by JosephSilber** [comment]
[I was never a fan of the `sole` method](https://twitter.com/joseph_silber/status/1350137212357246978), but I guess that ship has sailed.

**Comment by JosephSilber** [comment]
It's a pity the exceptions aren't shared (maybe down the line in 9.x). Then this:

https://github.com/laravel/framework/blob/277c2fbd0cebd2cb194807654d870f4040e288c0/src/Illuminate/Database/Concerns/BuildsQueries.php#L256-L266

could be changed to simply:

```php
return $this->take(2)->get($columns)->sole(); 
```

File: src/Illuminate/Collections/Collection.php (hunk 1 above)
Comment (ash-jc-allen) [reply src/Illuminate/Collections/Collection.php:1063]: Ahh okay, sorry I wasn't aware of that. I'll get them changed to be fully qualified

File: src/Illuminate/Collections/Enumerable.php (hunk 2 above)
Comment (ash-jc-allen) [reply src/Illuminate/Collections/Enumerable.php:835]: That makes sense thinking about it. I'll take this out and make a separate PR if this one gets merged

File: src/Illuminate/Collections/LazyCollection.php (hunk 3 above)
Comment (ash-jc-allen) [reply src/Illuminate/Collections/LazyCollection.php:1025]: I do agree that it's more complex. I was looking at some of the existing code for things like the ` first() ` method and just assumed that it had been done like that for a specific reason. I've not really used LazyCollections too much so I wasn't 100% sure on the best way to implement it. Thanks for the advice on this one

File: tests/Support/SupportCollectionTest.php (hunk 4 above)
Comment (ash-jc-allen) [reply tests/Support/SupportCollectionTest.php:74]: Yeah sure, I didn't realise those tests were there, so I'll add some to it

File: src/Illuminate/Collections/Traits/EnumeratesValues.php (hunk 5 above)
Comment (ash-jc-allen) [reply src/Illuminate/Collections/Traits/EnumeratesValues.php:469]: That makes sense. So do you suggest removed the ` soleWhere ` method completely?

File: src/Illuminate/Collections/Collection.php (hunk 6 above)
Comment (ash-jc-allen) [reply src/Illuminate/Collections/Collection.php:1067]: Ahh I must have missed that, cheers

**Comment by ash-jc-allen** [comment]
Thanks for the feedback on all of this @JosephSilber! I'll make all of the changes for it now :smile:

And I do agree that it'd be cool to have the shared functionality like in your suggestion.

**Comment by ash-jc-allen** [comment]
I've made the changes that you suggested @JosephSilber and from what I can see, they all seem to working. One thing thing I'm not too sure about though is the new test I've added. I have a feeling that I might not have done it right; apologies if that's the case.

File: tests/Support/SupportLazyCollectionIsLazyTest.php [hunk 7]
```diff
@@ -977,6 +977,17 @@ public function testSomeIsLazy()
...
+    public function testSoleIsLazy()
+    {
+        $data = $this->make([['a' => 1], ['a' => 2], ['a' => 3]]);
+
+        $this->assertEnumeratesCollection($data, 3, function ($collection) {
+            $collection->sole(function ($item) {
+                return $item['a'] === 2;
+            });
+        });
+    }
```
Comment (JosephSilber) [inline tests/Support/SupportLazyCollectionIsLazyTest.php:989]:
```suggestion
    public function testSoleIsLazy()
    {
        $this->assertEnumeratesOnce(function ($collection) {
            $collection->sole(function ($item) {
                return $item === 1;
            });
        });

        $this->assertEnumerates(4, function ($collection) {
            try {
                $collection->sole(function ($item) {
                    return $item % 2 === 0;
                });
            } catch (MultipleItemsFoundException $e) {
                //
            }
        });
    }
```

Be sure to also import `MultipleItemsFoundException`.

**Review by JosephSilber — COMMENTED** [review]
Looking better.

Seems like you forgot to add operator support to `sole`.

**Comment by JosephSilber** [comment]
> Seems like you forgot to add operator support to `sole`.

If it's not added now, we won't be able to add it until 9.x, since it would be a breaking change in the function signature.

File: tests/Support/SupportLazyCollectionIsLazyTest.php (hunk 7 above)
Comment (ash-jc-allen) [reply tests/Support/SupportLazyCollectionIsLazyTest.php]: Thanks for the suggestion again, I've just pushed up the new test changes :)

**Comment by ash-jc-allen** [comment]
@JosephSilber In your opinion, do you think it's something that should be added now? I can try and add it if so

**Comment by JosephSilber** [comment]
Would be nice.

If you can't figure it out, let me know, and I'll try to find a few minutes to do it for you 👍

**Comment by ash-jc-allen** [comment]
Sweet, I'll take a look at it now then. Thanks again for all the help on this one, I really appreciate it and it's giving me a chance to learn a bit more at the same time :smile:

**Comment by ash-jc-allen** [comment]
@JosephSilber I _think_ I've added the operator support, but please let me know if I've missed anything

File: src/Illuminate/Collections/Collection.php [hunk 8]
```diff
@@ -1050,6 +1052,37 @@ public function splitIn($numberOfGroups)
...
+
+            if ($items->count() > 1) {
+                throw new MultipleItemsFoundException;
+            }
+
+            return $items->first();
+        }
+
+        return $this->sole($this->operatorForWhere(...func_get_args()));
+    }
```
Comment (JosephSilber) [inline src/Illuminate/Collections/Collection.php:1084]: A ternary would be a little clearer, and allows dedenting the whole method:

```suggestion
    public function sole($key = null, $operator = null, $value = null)
    {
        $filter = func_num_args() > 1
            ? $this->operatorForWhere(...func_get_args())
            : $key;

        $items = $this->when($filter)->filter($filter);

        if ($items->isEmpty()) {
            throw new ItemNotFoundException;
        }

        if ($items->count() > 1) {
            throw new MultipleItemsFoundException;
        }

        return $items->first();
    }
```

File: src/Illuminate/Collections/LazyCollection.php [hunk 9]
```diff
@@ -1010,6 +1010,32 @@ public function split($numberOfGroups)
...
+            return $this
+                ->when($key)
+                ->filter($key)
+                ->take(2)
+                ->collect()
+                ->sole();
+        }
+
+        return $this->sole($this->operatorForWhere(...func_get_args()));
+    }
```
Comment (JosephSilber) [inline src/Illuminate/Collections/LazyCollection.php:1037]: Same here:

```suggestion
    public function sole($key = null, $operator = null, $value = null)
    {
        $filter = func_num_args() > 1
            ? $this->operatorForWhere(...func_get_args())
            : $key;
            
        return $this
            ->when($filter)
            ->filter($filter)
            ->take(2)
            ->collect()
            ->sole();
    }
```

**Comment by ash-jc-allen** [comment]
@JosephSilber Ah yeah, the ternary makes it look much cleaner, thanks! :smile:

**Review by JosephSilber — APPROVED** [review]
👍

File: src/Illuminate/Collections/Collection.php [hunk 10]
```diff
@@ -1050,6 +1052,37 @@ public function splitIn($numberOfGroups)
...
+    public function sole($key = null, $operator = null, $value = null)
+    {
+        $filter = func_num_args() > 1
+            ? $this->operatorForWhere(...func_get_args())
+            : $key;
+
+        $items = $this->when($filter)->filter($filter);
+
+        if ($items->isEmpty()) {
+            throw new ItemNotFoundException;
```
Comment (joelharkes) [inline src/Illuminate/Collections/Collection.php:1076]: would be nice if the query builder etc would throw the exact same exceptions when calling sole()

Comment (JosephSilber) [reply src/Illuminate/Collections/Collection.php:1076]: https://github.com/laravel/framework/pull/37034#issuecomment-822490782

**Comment by driesvints** [comment]
This PR apparently introduced a new `Illuminate\Collections` namespace which doesn't exists. It's not autoloaded. Only the `Illuminate\Support` namespace exists. I'll send a PR to fix this.

**Comment by ash-jc-allen** [comment]
@driesvints Sorry, my bad here! When I made the exceptions, I must have assumed at the time that the namespace was the same as the file path. I should have double-checked this before pushing 😔

**Comment by driesvints** [comment]
@ash-jc-allen don't worry about it. I've sent in a PR to fix this: https://github.com/laravel/framework/pull/37034
